[webkitscmpy] Read and write GitHub PR comments

JonWBedard opened the pull request:
#### 373c9a4b07136b26e325a3590e4227228ec23f65
<pre>
[webkitscmpy] Read and write GitHub PR comments
<a href="https://bugs.webkit.org/show_bug.cgi?id=266516">https://bugs.webkit.org/show_bug.cgi?id=266516</a>
<a href="https://rdar.apple.com/119739688">rdar://119739688</a>

Reviewed by Elliott Williams.

Provide Python functions to both write and read in-line PR comments.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.request): Mock endpoints to write and read PR comments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/pull_request.py:
(PullRequest.review): Add &apos;diff_comments&apos; argument.
(PullRequest.diff): Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.PRGenerator.review): Add currently unused &apos;diff_comments&apos; argument.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator._diff_comments): Return a dictionary of diff comments for this PR.
(GitHub.PRGenerator._make_comment): Construct a comment on this PR, passing the specified
kwargs directly to GitHub in the request body.
(GitHub.PRGenerator.review): Allow user to add comments to a specific file in
this PR, or to a specific line in a specific file.
(GitHub.PRGenerator.diff): Return the diff proposed by this pull-request, with
embedded comments if requested.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/scm.py:
(Scm.PRGenerator.diff): Added.
(Scm.insert_diff_comments): Function which inserts inline diff comments from an
SCM provider into a diff as if they were conflicts.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestNetworkPullRequestGitHub.webserver):
(TestNetworkPullRequestGitHub.test_diff):
(TestNetworkPullRequestGitHub.test_no_comments):
(TestNetworkPullRequestGitHub.test_comment):
(TestNetworkPullRequestGitHub.test_file_comment):
(TestNetworkPullRequestGitHub.test_comment_reply):

Canonical link: <a href="https://commits.webkit.org/277752@main">https://commits.webkit.org/277752@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1057b31d3d9d37bbdc0f80962880d0a8a83e4b2e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48469 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27681 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51157 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44534 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50775 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25204 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49051 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/25362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/20748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/48326 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/22845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6526 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53062 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/23516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/48500 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/24781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/25586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6900 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/24504 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -319,6 +365,40 @@ def review(self, pull_request, comment=None, approve=None):
+                        existing_ids = existing_comment_ids.get(file, {}).get(position, [])
+                        if existing_ids:
+                            is_successful &= self._make_comment(pull_request, dict(
+                                in_reply_to=existing_ids[-1],
```
It's possible to have multiple, distinct comments which begin on the same line in GitHub's UI, so I wonder if this could lead to replies being sent to the wrong comment thread in some cases.

JonWBedard replied:
This is a good point, and this is actually worse in BitBucket (because you can reply to replies in BitBucket). I'm going to add a FIXME for now. Addressing this won't be simple, we have to carry around comment IDs with us, or we have to include some sort of "reply to" logic. The real issue is how our callers do this, so I'm inclined to postpone resolving this problem until we understand how our callers will work better. There is also the fact that, in practice, users very rarely comment on a PR line without replying to the existing comment, so this is an edge case we're unlikely to be bothered by in practice.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -356,7 +436,7 @@ def review(self, pull_request, comment=None, approve=None):
                 ))
 
-            return pull_request
+            return pull_request if is_successful else None
```
nit: since None is implicit, you could instead write
```suggestion
            if is_successful:
                return pull_request
```

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/scm.py:
```diff
@@ -82,6 +85,25 @@ def from_url(cls, url, contributors=None, classifier=None):
+                file = line.split('/', 1)[-1]
+                count = -1
+            yield line
+            comments_on = comments.get(file, {}).get(None if count < 0 else count, [])
+            if comments_on:
+                yield '>>>>'
+                for comment in comments_on:
+                    for line in comment.splitlines():
+                        yield line
+                yield '<<<<'
+            count += 1
```
We can certainly play around with the formatting, but I wonder if, instead of interspersing comments with what look like conflict markers, it would be better to indenting all the code rightward one level, and leave comments unindented? That ought to help them visually stand out, without adding line noise that would confuse a screen reader.

JonWBedard replied:
The good news is that this is indeed shared code, I too, am unsure that `<<<<` is the right approach.

One of the challenges here is that we're not actually interested in comments "visually" standing out, we're interested in them being easy for screen readers to parse. I did consult folks using screen readers, and `<<<<` was the implementation we settled on. I'm not certain if indenting all the code is better or worse for screen reader users, I intend to follow up with screen reader users once we have a PR which can download a diff with comments.

JonWBedard committed:
[webkitscmpy] Read and write GitHub PR comments
https://bugs.webkit.org/show_bug.cgi?id=266516
rdar://119739688

Reviewed by Elliott Williams.

Provide Python functions to both write and read in-line PR comments.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.request): Mock endpoints to write and read PR comments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/pull_request.py:
(PullRequest.review): Add 'diff_comments' argument.
(PullRequest.diff): Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.PRGenerator.review): Add currently unused 'diff_comments' argument.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator._diff_comments): Return a dictionary of diff comments for this PR.
(GitHub.PRGenerator._make_comment): Construct a comment on this PR, passing the specified
kwargs directly to GitHub in the request body.
(GitHub.PRGenerator.review): Allow user to add comments to a specific file in
this PR, or to a specific line in a specific file.
(GitHub.PRGenerator.diff): Return the diff proposed by this pull-request, with
embedded comments if requested.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/scm.py:
(Scm.PRGenerator.diff): Added.
(Scm.insert_diff_comments): Function which inserts inline diff comments from an
SCM provider into a diff as if they were conflicts.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestNetworkPullRequestGitHub.webserver):
(TestNetworkPullRequestGitHub.test_diff):
(TestNetworkPullRequestGitHub.test_no_comments):
(TestNetworkPullRequestGitHub.test_comment):
(TestNetworkPullRequestGitHub.test_file_comment):
(TestNetworkPullRequestGitHub.test_comment_reply):

Canonical link: https://commits.webkit.org/277752@main
